src/node_messaging.cc

```diff
@@ -0,0 +1,831 @@
+      main_message_buf_.len,
+      &delegate);
+  delegate.deserializer = &deserializer;
+
+  // Attach all transfered (Shared)ArrayBuffers to their new Isolate.
+  for (uint32_t i = 0; i < array_buffer_contents_.size(); ++i) {
+    Local<ArrayBuffer> ab =
+        ArrayBuffer::New(env->isolate(),
+                         array_buffer_contents_[i].base,
+                         array_buffer_contents_[i].len,
+                         ArrayBufferCreationMode::kInternalized);
+    array_buffer_contents_[i].base = nullptr;
+    deserializer.TransferArrayBuffer(i, ab);
+  }
+  array_buffer_contents_.clear();
+
+  for (uint32_t i = 0; i < shared_array_buffers_.size(); ++i) {
+    Local<SharedArrayBuffer> sab;
+    if (!shared_array_buffers_[i]->GetSharedArrayBuffer(env, context)
+            .ToLocal(&sab))
+      return MaybeLocal<Value>();
+    deserializer.TransferSharedArrayBuffer(i, sab);
+  }
+  shared_array_buffers_.clear();
+
+  if (deserializer.ReadHeader(context).IsNothing())
+    return MaybeLocal<Value>();
+  return handle_scope.Escape(
+      deserializer.ReadValue(context).FromMaybe(Local<Value>()));
+}
+
+void Message::AddSharedArrayBuffer(ExternalSABReference reference) {
+  shared_array_buffers_.push_back(reference);
+}
+
+void Message::AddMessagePort(std::unique_ptr<MessagePortData>&& data) {
+  message_ports_.emplace_back(std::move(data));
+}
+
+namespace {
+
+class SerializerDelegate : public ValueSerializer::Delegate {
+ public:
+  SerializerDelegate(Environment* env, Local<Context> context, Message* m)
+      : env_(env), context_(context), msg_(m) {}
+
+  void ThrowDataCloneError(Local<String> message) override {
+    env_->isolate()->ThrowException(Exception::Error(message));
+  }
+
+  Maybe<bool> WriteHostObject(Isolate* isolate, Local<Object> object) override {
+    if (env_->message_port_constructor_template()->HasInstance(object)) {
+      return WriteMessagePort(Unwrap<MessagePort>(object));
+    }
+
+    env_->ThrowError("Cannot serialize unknown type of host object");
+    return Nothing<bool>();
+  }
+
+  Maybe<uint32_t> GetSharedArrayBufferId(
+      Isolate* isolate, Local<SharedArrayBuffer> shared_array_buffer) override {
+    uint32_t i;
+    for (i = 0; i < seen_shared_array_buffers_.size(); ++i) {
+      if (seen_shared_array_buffers_[i] == shared_array_buffer)
+        return Just(i);
+    }
+
+    ExternalSABReference reference(
+        ReferenceCountedSAB::ForIncomingSharedArrayBuffer(env_,
+                                                          context_,
+                                                          shared_array_buffer));
+    if (!reference) {
+      return Nothing<uint32_t>();
+    }
+    seen_shared_array_buffers_.push_back(shared_array_buffer);
+    msg_->AddSharedArrayBuffer(reference);
+    return Just(i);
+  }
+
+  void Finish() {
+    for (MessagePort* port : ports_) {
+      port->Close();
+      msg_->AddMessagePort(std::move(port->Detach()));
```

Comment:
Is the std::move redundant if MessagePort::Detach already returns a std::move'd smart pointer?

Reply:
Clang complains:

```
../src/node_messaging.cc:325:28: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
      msg_->AddMessagePort(std::move(port->Detach()));
                           ^
../src/node_messaging.cc:325:28: note: remove std::move call here
      msg_->AddMessagePort(std::move(port->Detach()));
                           ^~~~~~~~~~              ~
```

Will fix.